### PR TITLE
Add EnergyDensity function and tags for the scalar wave system

### PR DESCRIPTION
**Pull request by MarloMo**
## Proposed changes

Add in the `EnergyDensity.cpp` and `EnergyDensity.hpp` that provides the `energy_density` function that calculates the energy density over the domain of a scalar wave system. Add the `EnergyDensityCompute` to `Tags.hpp` that calculates the energy density using the `energy_density` function. 

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp]: Unused.

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp]: Include `"DataStructures/Tensor/TypeAliases.hpp"` instead.

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp]: Replace with a forward declaration of `gsl::not_null`.

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp]: Forward declare `DataVector`.

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp:32]: Include `<cstddef>`.

File: src/Evolution/Systems/ScalarWave/EnergyDensity.cpp
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.cpp]: Original change I was going to suggest, which I'm leaving because it has useful explanation:
`square(get(pi)) + get(dot_product(phi, phi))`.

Include `"Utilities/ConstantExpressions.hpp"` for `square`, replace the include of `Magnitude.hpp` with `DotProduct.hpp`.

`magnitude` just calls `dot_product` and then takes the square root, so this avoids taking the square root and then squaring it.  The `square` in the first term is just readability/style.

Better change:

Use the return-by-reference form of `dot_product` to assign phi^2 to `result`, then add square(pi), and then multiply the whole thing by 0.5.  This avoids allocating memory for the return value of `dot_product`/`magnitude`.  Change includes same way as the original version of the suggestion.

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp:22]: This is unmatched.  Remove it.

File: src/Evolution/Systems/ScalarWave/Tags.hpp
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/Tags.hpp]: Remove.

**Review by wthrowe — CHANGES_REQUESTED** [review]
This needs tests.  (For both the function and the tags.)

In the future, please don't reformat entire files if you only change a few lines.

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp:36]:
```cpp
template <size_t SpatialDim>
Scalar<DataVector> energy_density(
  const Scalar<DataVector>& pi,
    const tnsr::i<DataVector, SpatialDim, Frame::Inertial>& phi) noexcept;
```

in the implementation (cpp file)
```cpp
 {
  Scalar<DataVector> result{get(pi).size()};
  energy_density(make_not_null(&result), pi, phi);
  return result;
}
```

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp:28]: can you please add a blank line before the `\f{align*}` As in:
```cpp
 * Below is the fucntion used to calculate the energy density.
 * 
 * \f{align*}
```

also typo: `fucntion` -> `function`

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp:22]: this needs a closing `/// @}` after the _second_ function. E.g.
```cpp
/// @{
/// My very useful comment
void foo(gsl::not_null<Scalar<DataVector>*> result, Scalar<DataVector> input) noexcept;

Scalar<DataVector> foo(Scalar<DataVector> input) noexcept;
/// @}
```

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp]: in forward declarations you don't need `const` on variables that are taken by value. As in:

```cpp
void foo1(double a);
void foo2(const double& a);
void foo3(double& a);
```
the `&` makes the `const` important because without the const, you could modify the variable `a` in the callers scope. That is,
```cpp
const double b = 1.2;
foo1(b); // will never modify b because `b` is copied
foo2(b); // will never modify `b` because `b` is a `const &`
foo3(b); // might modify `b`, you don't know unless you look inside `foo3`
```

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp:5]: includes should be:
```cpp
#include <cstddef>

#include "DataStructures/Tensor/TypeAliases.hpp"

/// \cond
class DataVector;
namespace gsl {
template <typename T>
class not_null;
}  // namespace gsl
/// \endcond
```

File: src/Evolution/Systems/ScalarWave/EnergyDensity.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.cpp]: need `#include "DataStructures/DataVector.hpp`

File: src/Evolution/Systems/ScalarWave/EnergyDensity.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.cpp:11]: `#include "Utilities/ConstantExpressions.hpp`

File: src/Evolution/Systems/ScalarWave/EnergyDensity.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.cpp]: `square(get(magnitude(phi)))`

File: src/Evolution/Systems/ScalarWave/Tags.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/Tags.hpp]: (I think) you can use this to combine the documentation for the `EnergyDensity` and `EnergyDensityCompute` by moving it up to line 108 and then closing it after `EnergyDensityCompute`

File: src/Evolution/Systems/ScalarWave/Tags.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/Tags.hpp]: blank line before and `fucntion` typo

File: src/Evolution/Systems/ScalarWave/Tags.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/Tags.hpp]: you'll need to do a `static_cast` and turn this into something similar to https://github.com/sxs-collaboration/spectre/blob/08973d05a7262714ec71b006b40fc88b4f1c9058/src/PointwiseFunctions/GeneralRelativity/Lapse.hpp#L66

File: src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp
Comment (nilsdeppe) [inline src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp]: I would split this commit into 3:
1. add `energy_density` function and tests
2. add `EnergyDensity` tag and test
3. `EnergyDensityCompute` tags and tests (add into `EnergyDensity.hpp` instead of `Tags.hpp`)
4. update executable

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: tests/Unit/Evolution/Systems/ScalarWave/Test_EnergyDensity.cpp
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/ScalarWave/Test_EnergyDensity.cpp]: I don't find the name `sdv` to be very descriptive. Can you pass the `used_for_size` directly into your `test_energy_density` function by having it accept a `DataVector&` instead of a `Scalar<DataVector>&`? I thought `check_with_random_values` would accept a `DataVector` as its argument. Doing so would allow you to eliminate the `sdv` variable completely. You also don't need all the random number generation and the `DataVector` named `used_for_size` should be of size `5`. I think all you need is:

```cpp
const DataVector used_for_size(5);
test_energy_density<1>(used_for_size);
test_energy_density<2>(used_for_size);
test_energy_density<3>(used_for_size);
```
Note that I also added tests for 2d and 3d in the snippet

File: tests/Unit/Evolution/Systems/ScalarWave/Test_EnergyDensity.cpp
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/ScalarWave/Test_EnergyDensity.cpp]: you're adding a test for the _simple_ tag, which should go in `Test_Tags.cpp`

File: tests/Unit/Evolution/Systems/ScalarWave/Test_EnergyDensity.cpp
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/ScalarWave/Test_EnergyDensity.cpp:44]: please add a blank line between the functions. It makes the code easier to read :)

File: tests/Unit/Evolution/Systems/ScalarWave/Test_EnergyDensity.cpp
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/ScalarWave/Test_EnergyDensity.cpp:8]: you don't use `cmath`

you don't use `DataBox.hpp` or `TMPL.hpp`

you won't be using many of the others after you've updated the code :)

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp:46]: need to `#include "DataStructures/DataBox/Tag.hpp` for `db::ComputeTag`

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp:47]: include `Utilities/TMPL.hpp`

File: src/Evolution/Systems/ScalarWave/EnergyDensity.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.cpp]: Please move this change into the right commit

File: src/Evolution/Systems/ScalarWave/TagsDeclarations.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/TagsDeclarations.hpp:35]: Please move this change into the right commit

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/ScalarWave/EnergyDensity.cpp
Comment (MarloMo) [reply src/Evolution/Systems/ScalarWave/EnergyDensity.cpp]: Would this be the correct way to write the dot_product version? @wthrowe , @nilsdeppe 

``` 
auto sqr_phi = get(dot_product(phi, phi));
get(*result) = 0.5 * (square(get(pi)) + sqr_phi);
```

Comment (nilsdeppe) [reply src/Evolution/Systems/ScalarWave/EnergyDensity.cpp]: I believe what @wthrowe has in mind is:
```cpp
dot_product(result, phi, phi);
get(*result) += square(get(pi));
get(*result) *= 0.5;
```

Comment (wthrowe) [reply src/Evolution/Systems/ScalarWave/EnergyDensity.cpp]: Yes, that's what I was thinking.

File: tests/Unit/Evolution/Systems/ScalarWave/Test_EnergyDensity.cpp
Comment (nilsdeppe) [reply tests/Unit/Evolution/Systems/ScalarWave/Test_EnergyDensity.cpp]: you still need to remove the `test_compute_item_in_databox` from the commit "Add EnergyDensity simple tag and test"

File: src/Evolution/Systems/ScalarWave/Tags.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/Tags.hpp:109]: this change should be in the commit that adds the tag

File: src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp
Comment (nilsdeppe) [inline src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp:57]: In the future can you please make sure you only format the lines you're changing? The commit message makes this sound like a very short but there is a lot going on in this commit. Try to make the life of the reviewer as easy as possible :)

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp
Comment (MarloMo) [reply src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp:57]: @wthrowe , @nilsdeppe , we are ready for a final review.

File: tests/Unit/Evolution/Systems/ScalarWave/Test_EnergyDensity.cpp
Comment (wthrowe) [inline tests/Unit/Evolution/Systems/ScalarWave/Test_EnergyDensity.cpp:18]: I think `Evolution/Systems/ScalarWave/Tags.hpp`, `Helpers/DataStructures/DataBox/TestHelpers.hpp`, and `Helpers/DataStructures/MakeWithRandomValues.hpp` are unused. --- OK, just move them to the compute tag commit.

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp:8]: Unused. --- Actually, just in the wrong commit.  Move to the compute tag commit.

File: src/Evolution/Systems/ScalarWave/Tags.hpp
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/Tags.hpp]: You don't need this include.

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp]: This isn't working, probably because the tag is in a different namespace.

File: src/Evolution/Systems/ScalarWave/Tags.hpp
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/Tags.hpp]: I think the changes to this file were squashed into the wrong commit.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp:42]: following up on @wthrowe's comment about the doxygen grouping. Here you want:
```cpp
     const tnsr::i<DataVector, SpatialDim, Frame::Inertial>& phi) noexcept;
/// @}

namespace Tags {
/// \brief Computes the energy density using ScalarWave::energy_density()
template <size_t SpatialDim>
struct EnergyDensityCompute...
```
and then remove the `/// @}` on line 56 below

**Comment by wthrowe** [comment]
The doxygen changes look good.  I find all the code moving between commits too confusing to review as fixups, so I'll look that over after they are squashed.

**Comment by MarloMo** [comment]
@wthrowe squash complete

**Review by nilsdeppe — DISMISSED** [review]
I don't need to re-approve if @wthrowe has any more changes he'd like to see before this gets merged :)

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (wthrowe) [reply src/Evolution/Systems/ScalarWave/EnergyDensity.hpp:46]: This got lost in the shifting of changes between commits.

**Review by wthrowe — CHANGES_REQUESTED** [review]
Not a problem unless you want to change it, but in many of your commits you are listed as "marlo \<marlo@wheeler.caltech.edu\>" instead of your GitHub account.  You can change the values used when you commit with `git config --global user.name "Your Name"` and `git config --global user.email "your@address"`, but I don't know an easy way to change the values used for already-existing commits.

Edit: formatting

**Comment by wthrowe** [comment]
Squash.

File: src/Evolution/Systems/ScalarWave/EnergyDensity.hpp
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/EnergyDensity.hpp]: Oops, I missed that: Missing `"`.

**Comment by wthrowe** [comment]
Squash.  And rebase, apparently.

**Review by wthrowe — APPROVED** [review]

**Pull request merged**
